src/org/rascalmpl/util/maven/Project.java

```diff
@@ -0,0 +1,274 @@
+    }
+
+    public IList getMessages() {
+        return messages;
+    }
+
+    public @Nullable String getLicense() {
+        throw new UnsupportedOperationException("Not implemented yet");
+    }
+
+    private Project(ArtifactCoordinate coordinate, @Nullable ArtifactCoordinate parentCoordinate,
+        List<Dependency> dependencies, IList messages, boolean errors) {
+        this.coordinate = coordinate;
+        this.parentCoordinate = parentCoordinate;
+        this.dependencies = dependencies;
+        this.messages = messages;
+        this.errors = errors;
+    }
+    /**
+     * We have to have a file name to a pom.xml in a project, it does not work for .pom files in a maven repository.
+     * Sometimes you want to configure where the jars should be stored, in most cases use the other overload
+     * @param rootPath of the maven repo
+     * @return
+     */
+    public static Project parseProjectPom(Path pomFile, Path mavenRoot) {
+        var modelBuilder = new DefaultModelBuilderFactory().newInstance();
+        var request = new DefaultModelBuildingRequest()
+            .setPomFile(pomFile.toFile())
+            .setValidationLevel(ModelBuildingRequest.VALIDATION_LEVEL_MAVEN_3_0); // TODO: figure out if we need this
+        var httpClient = HttpClient.newBuilder()
+            .version(Version.HTTP_2) // upgrade where possible
+            .connectTimeout(Duration.ofSeconds(10)) // don't wait longer than 10s to connect to a repo
+            .build();
+        var context = new CurrentResolution(modelBuilder, new HashMap<>(), makeLocation(pomFile), httpClient, mavenRoot);
+        return build(request, context, true);
+    }
+
+    /**
+     * We have to have a file name to a pom.xml in a project, it does not work for .pom files in a maven repository.
+     */
+    public static Project parseProjectPom(Path pomFile) {
+        return parseProjectPom(pomFile, Util.mavenRepository());
+    }
+
+    private static Project build(ModelBuildingRequest request, CurrentResolution context, boolean isRoot) {
+        try {
+            request.setModelResolver(context.resolver)
+                .setSystemProperties(System.getProperties());
+            // TODO: maybe 2 phase bulding is nice ?
+            return translate(context.builder.build(request), context, isRoot);
+        } catch (ModelBuildingException be) {
+            return buildError(be, context, isRoot);
+        }
+    }
+
+
+    static Project parseRepositoryPom(ModelSource resolvedEntry, CurrentResolution context) {
+        var request = new DefaultModelBuildingRequest().setModelSource(resolvedEntry);
+        URI loc;
+        if (resolvedEntry instanceof ModelSource2) {
+            loc = ((ModelSource2)resolvedEntry).getLocationURI();
+        }
+        else {
+            try {
+                loc = new URI(resolvedEntry.getLocation());
+            }
+            catch (URISyntaxException e) {
+                loc = URIUtil.unknownLocation().getURI();
```

Comment:
This should probably not go unnoticed. Why not throw a runtime error?

Reply:
Well, this is only for the location of the pom.xml in an error message.

Right now there's some more stuff I would like to get right, such as proper offsets (there is a todo about that). But this exceptional case I'm fine with ignoring. It can only occur when a pom came from a non-file (or even non-uri) location.